dhcpv4/bsdp.go

```diff
@@ -0,0 +1,328 @@
+// +build darwin
+
+package dhcpv4
+
+// Implements Apple's netboot protocol BSDP (Boot Service Discovery Protocol).
+// Canonical implementation is defined here:
+// http://opensource.apple.com/source/bootp/bootp-198.1/Documentation/BSDP.doc
+
+import (
+	"encoding/binary"
+	"fmt"
+	"syscall"
+)
+
+// Options (occur as sub-options of DHCP option 43).
+const (
+	BSDPOptionMessageType OptionCode = iota + 1
+	BSDPOptionVersion
+	BSDPOptionServerIdentifier
+	BSDPOptionServerPriority
+	BSDPOptionReplyPort
+	BSDPOptionBootImageListPath // Not used
+	BSDPOptionDefaultBootImageID
+	BSDPOptionSelectedBootImageID
+	BSDPOptionBootImageList
+	BSDPOptionNetboot1_0Firmware
+	BSDPOptionBootImageAttributesFilterList
+)
+
+// Versions (seen so far)
+var (
+	BSDPVersion1_0 = []byte{1, 0}
+	BSDPVersion1_1 = []byte{1, 1}
+)
+
+// BSDP message types
+const (
+	BSDPMessageTypeList byte = iota + 1
+	BSDPMessageTypeSelect
+	BSDPMessageTypeFailed
+)
+
+// Boot image kinds
+const (
+	BSDPBootImageMacOS9 byte = iota
+	BSDPBootImageMacOSX
+	BSDPBootImageMacOSXServer
+	BSDPBootImageHardwareDiagnostics
+	// 0x4 - 0x7f are reserved for future use.
+)
+
+// BootImageID describes a boot image ID - whether it's an install image and
+// what kind of boot image (e.g. OS 9, macOS, hardware diagnostics)
+type BootImageID struct {
+	isInstall bool
+	imageKind byte
+	index     uint16
+}
+
+// toBytes serializes a BootImageID to network-order bytes.
+func (b BootImageID) toBytes() (bytes []byte) {
+	bytes = make([]byte, 4)
+	// Attributes.
+	if b.isInstall {
+		bytes[0] |= 0x80
+	}
+	bytes[0] |= b.imageKind
+
+	// Index
+	binary.BigEndian.PutUint16(bytes[2:], b.index)
+	return
+}
+
+// BootImageIDFromBytes deserializes a collection of 4 bytes to a BootImageID.
+func bootImageIDFromBytes(bytes []byte) BootImageID {
```

Comment:
`golint` here will likely complain about the mismatch between function name and first word in the comment. By the way this seems a good candidate for a public function too, any reason to keep it private?

Reply:
Mmm, no real reason. I'm still learning what bits to make public and what to make private.